enhancements/cluster-logging/forwarding-json-structured-logs.md

```diff
@@ -0,0 +1,358 @@
+
+#### I want to access fields in all JSON log messages.
+
+I am using `fluentForward` and I want to be able to access any field in my JSON log messages.
+I used to use the [`MERGE_JSON_LOG`][merge] feature to do this, but that was removed.
+
+Example 1: I want non-JSON application messages to be discarded. JSON messages are forwarded in the `structured_message` field.
+```yaml
+outputs:
+- name: jsonOut
+  structured:
+    discard: unstructured
+pipelines:
+- inputRefs: [ application ]
+  outputRefs: [ jsonOut ]
+```
+
+Example 2: I want both JSON and non-JSON messages forwarded. JSON in the `structured_message` field, non-JSON in the `message` field.
+```yaml
+outputs:
+- name: jsonOut
+  structured:
+    discard: none
+pipelines:
+- inputRefs: [ application ]
+  outputRefs: [ jsonOut ]
+```
+
+#### I want to  forward structured and unstructured logs from the same input to different outputs
+
+```yaml
+outputs:
+- name: structuredOut
+  structured:
+    discard: unstructured
+- name: unstructuredOut
+  structured:
+    discard: structured
+pipelines:
+- inputRefs: [ application ]
+  outputRefs: [ structuredOut, unstructuredOut ]
+```
+
+### Proposal details
+
+#### Data model changes
+
+Two new top-level fields added to the internal [data model][model]:
+* `structured_message` (type object): Original structured JSON log message.
+* `structured_format` (type string): Format name from `input.structured.format`.
+
+Note: `structured_format` is an arbitrary name assigned by the user, it has no meaning to the forwarder.
+
+The following rules apply to forwarded records:
+* At least one of `message` or `structured_message` MUST be present. They MAY both be present.
+* `message` and `structured_message` MAY both be present, if so `message` MUST be the JSON quoted-string form of the value of `structured_message`.
+* `structured_format` MAY be present if `structured_message` is present, and MUST NOT be present if `structured_message` is missing.
+
+Interpretation of the record:
+* `structured_format` is present: it indicates the format of `structured_message`
+* `structured_format` is missing:
+  * `structured_message` is present it: it is JSON, but of an unknown format.
+  * `structured_message` is missing: `message` is the original message of unknown encoding.
+* If both `structured_message` and `message` are present they are redundant copies of the same data. May be useful in upgrade transitions.
+
+#### ClusterLogForwarder configuration
+
+New field `input.stuctured` with sub-field
+* `format`: Name to forward as value of `structured_format`. If missing, no `structured_format` field is forwarded.
+
+New field `output.structured` with sub-fields:
+* `discard` (enum: Structured, Unstructured) discard either structured or unstructured messages. If missing do not discard anything.
+* `unstructured`: enum indicating treatment of unstructured `message` field.
```

Comment:
Not sure I completely understand the necessity of this field when you already have `discard`. Can you elaborate.

Reply:
Yep, see new proposal, I think it's more self-explanatory now.